LFS-349: Add a confirmation dialog prior to uninstalling vocabulary

veronikaslc opened the pull request:
LFS-350: Add Close button to the right of Uninstall in the vocabulary 's About dialog
LFS-384: Simple confirmation dialog
LFS-385: List variables linked to this vocabulary

Commit:
LFS-350: Add Close button to the right of Uninstall in the vocabulary's About dialog

Commit:
LFS-350: Add Close button to the right of Uninstall in the vocabulary's About dialog
Removed close icon

Commit:
LFS-384: Simple confirmation dialog

Commit:
LFS-385: List variables linked to this vocabulary

sashaandjic commented:
When uninstalling vocabulary that is not in use, the dialog appears as expected:
![image](https://user-images.githubusercontent.com/23243244/79897930-64784680-83d8-11ea-8541-4aa6bdb32ee8.png)
Not sure about the color of the Cancel bttn. Users may think is it disabled.
When trying to uninstall vocabulary already in use the dialog appears listing HPO terms that it found on the forst attempt. In the meantime I replaced those HPO terms, and added another form with term, however none of them were listed. I am not eve sure if we need to display the terms at all. I believe it is enough just to say that the vocabulary is used in forms.  
![image](https://user-images.githubusercontent.com/23243244/79898193-d6e92680-83d8-11ea-8c1e-03b59ac31836.png)

The VM is running at http://172.20.4.94:8080/

sdumitriu commented on modules/vocabularies/src/main/frontend/src/vocabularyAction.jsx:
```diff
@@ -74,6 +80,64 @@ const useStyles = makeStyles(theme => ({
+  let fetchQuestionnaires = () => {
+    if (questionnaires.length === 0) {
+      // Send a fetch request to determine the questionnaires available
+      fetch('/query?query=' + encodeURIComponent('select * from [lfs:Questionnaire]'))
```
If you don't use a more complicated query, you can just get `/Questionnaires.paginate`.

By default, at most 10 questionnaires are returned, which right now is OK, but once we have more than that you may miss some usages of the vocabulary. You should request a high `limit`.

I can't remember, why don't you query just the questionnaires that actually use this vocabulary?

```suggestion
      const query = `select n.* from [lfs:Questionnaire] as n inner join [lfs:Question] as q on isdescendantnode(q, n) where q.sourceVocabulary='${props.acronym}'`;
      fetch(`/query?query=${encodeURIComponent(query)}&limit=100`)
```

If there are more than 100 questionnaires using this vocabulary, it probably wouldn't fit anyway. If we select just the questionnaires that do use it, we could even set a limit of 5 and say "... and 20 more others".

veronikaslc replied:
I can not query just the questionnaires because I need to display both questions and questionnaires names they belong to.

sdumitriu commented on modules/vocabularies/src/main/frontend/src/vocabularyAction.jsx:
```diff
@@ -74,6 +80,64 @@ const useStyles = makeStyles(theme => ({
+        .then((data) => {
+          aggregatedQuestions = aggregatedQuestions.concat(getVocabularyQuestions(data, questionnaire.title));
+
+          if (++i == questionnairesData.length) {
```
If one of them errors out, we'll never get here. This part should happen in a `finally` callback.

marta- commented:
> once I deleted all HPO terms from all forms, I was still getting the same warning listing HPO terms that were in use when I first attempted to uninstall HPO.

It shouldn't matter if there are any HPO terms actually entered in any filled out forms. It only matters that a question expects HPO terms as answers. If HPO terms are expected (specified in the question definition in the questionnaire) but the vocabulary is not installed so the suggestions are unavailable, users will not be able to properly answer that question.

marta- commented:
> Not sure about the color of the Cancel bttn. Users may think is it disabled.

Agreed. This is caused by custom styling applied to all buttons in the vocabulary administration page. That makes all buttons have white text.

@veronikaslc this wasn't introduced by you, can you rework the existing style so that:
* only buttons inside the table get lowercased, while dialog actions preserve their default look
* only change the text color whenever a custom background color is specified (e.g. red + white for uninstall) rather than make all buttons white text
* try to use theme colors (error, success,...) Instead of hardcoded colors.

Additionally, with the changes above, making the close button blue on focus becomes unnecessary (and off-theme) and that line should be removed. The clise button should have the style of a material button with color=default.

veronikaslc committed:
LFS-385: List variables linked to this vocabulary

Query just the questionnaires that actually use the particular vocabulary

Co-Authored-By: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

Commit:
LFS-385: List variables linked to this vocabulary
Addressed codereview comments

veronikaslc committed:
LFS-350: Add Close button to the right of Uninstall in the vocabulary 's About dialog

Removed special styling for close button.

Commit:
LFS-385: List variables linked to this vocabulary
Corrected styles and fixed small bug

sdumitriu commented on modules/vocabularies/src/main/frontend/src/vocabularyAction.jsx:
```diff
@@ -74,6 +80,69 @@ const useStyles = makeStyles(theme => ({
+      // Send a fetch request to determine the questionnaires available
+      const query = `select n.* from [lfs:Questionnaire] as n inner join [lfs:Question] as q on isdescendantnode(q, n) where q.sourceVocabulary='${props.acronym}'`;
+      fetch(`/query?query=${encodeURIComponent(query)}&limit=100`)
+        .then((response) => response.ok ? response.json() : Promise.reject(response))
```
A problem with fetches is that in case the user is no longer authenticated, the fetch will return the login page, and this will silently fail because the "JSON" is invalid, since it is actually an HTML. This can be solved here by adding a catch or finally.

marta- committed:
[misc] Replaced tabs with spaces

marta- committed:
LFS-349: Add a confirmation dialog prior to uninstalling vocabulary
- renamed the css class for vocabulary actions to be more specific
- used contrast colors defined in the theme.pallette instead of hardcoded "white"
- fixed the style of the buttons in the "uninstall" confirmation dialog

Commit:
LFS-349: Add a confirmation dialog prior to uninstalling vocabulary
Added .catch() clause for catching possible errors in questionaires fetching

veronikaslc commented:
Ok, now I'm totally confused.

veronikaslc commented:
Re-posting the last Marta's comment
>Please revert this change.

I should have given more details with my commit which contradicts the earlier review. Additionally the review text itself is ambiguous: by default look I just meant uppercased, but with the "right" custom color.

That said, your changes did not give the buttons in the About dialog the default style (uppercase) by using a more restrictive selector as we had discussed in the meeting (but I failed to specify afterwards on GitHub). However the confirmation dialog that appears fir Uninstall has fully default looking actions, which is inconsistent: we have one "uninstall" button red and capitalized, and in the dialog on top of it we have another uninstall button that is blue and uppercased.

To quickly make everything look consistent I just kept the capitalization and applied it to all vocabulary buttons, but also applied the custom color everywhere.

You make a modification to return the dialog buttons to uppercase style.

veronikaslc committed:
Merge branch 'LFS-349' into LFS-350

veronikaslc committed:
Merge pull request #191 from ccmbioinfo/LFS-350

LFS-350: Add Close button to the right of Uninstall in the vocabulary 's About dialog

veronikaslc commented:
Merge with [LFS-350 PR](https://github.com/ccmbioinfo/lfs/pull/191/) for the convenience of simultaneous testing.

sashaandjic reviewed (approved):
tested

![image](https://user-images.githubusercontent.com/23243244/80546881-c579be80-8984-11ea-95cb-1d58f9abe08f.png)

![image](https://user-images.githubusercontent.com/23243244/80544342-d6273600-897e-11ea-9fb9-7b440ed8d47a.png)
